cirrus/google_cloud/utils.py

```diff
@@ -230,6 +238,137 @@ def _get_string_to_sign(
+        service_account_creds (dict, optional): JSON keyfile dict for Google
+            Service Account (can be obtained by calling `get_access_key`)
+        requester_pays_user_project (optional): User's Google project for billing
+
+    Returns:
+        str: Completed signed URL
+    """
+
+    if service_account_creds:
+        creds = service_account.Credentials.from_service_account_info(
+            service_account_creds
+        )
+    else:
+        # Default creds
+        creds = service_account.Credentials.from_service_account_info(
+            config.GOOGLE_APPLICATION_CREDENTIALS
+        )
+    bucket_name = path_to_resource.split("/")[0]
+    object_name = "/".join(path_to_resource.split("/")[1:])
+    escaped_object_name = six_quote(six.ensure_binary(object_name), safe=b"/~")
+    canonical_uri = "/{}".format(escaped_object_name)
+
+    datetime_now = datetime.datetime.utcnow()
+    request_timestamp = datetime_now.strftime("%Y%m%dT%H%M%SZ")
+    datestamp = datetime_now.strftime("%Y%m%d")
+
+    client_email = creds.service_account_email
+    credential_scope = "{}/auto/storage/goog4_request".format(datestamp)
+    credential = "{}/{}".format(client_email, credential_scope)
+
+    if extension_headers is None:
+        extension_headers = dict()
+
+    host = (
+        "{}.storage.googleapis.com".format(bucket_name)
+        if bucket_name
+        else "storage.googleapis.com"
+    )
+    extension_headers["host"] = host
+
+    canonical_headers = ""
+    ordered_headers = collections.OrderedDict(sorted(extension_headers.items()))
+    for k, v in ordered_headers.items():
+        lower_k = str(k).lower()
+        strip_v = str(v).lower()
+        canonical_headers += "{}:{}\n".format(lower_k, strip_v)
+
+    signed_headers = ""
+    for k, _ in ordered_headers.items():
+        lower_k = str(k).lower()
+        signed_headers += "{};".format(lower_k)
+    signed_headers = signed_headers[:-1]  # remove trailing ';'
+
+    if canonical_query_params is None:
+        canonical_query_params = dict()
+    canonical_query_params["x-goog-algorithm"] = "GOOG4-RSA-SHA256"
+    canonical_query_params["x-goog-credential"] = credential
+    canonical_query_params["x-goog-date"] = request_timestamp
+    canonical_query_params["x-goog-expires"] = expires
+    canonical_query_params["x-goog-signedheaders"] = signed_headers
+
+    if requester_pays_user_project is not None:
+        canonical_query_params["userProject"] = requester_pays_user_project
+
+    canonical_query_string = ""
+    ordered_query_parameters = collections.OrderedDict(
+        sorted(canonical_query_params.items())
+    )
+    for k, v in ordered_query_parameters.items():
+        encoded_k = quote(str(k), safe="")
+        encoded_v = quote(str(v), safe="")
+        canonical_query_string += "{}={}&".format(encoded_k, encoded_v)
+    canonical_query_string = canonical_query_string[:-1]  # remove trailing '&'
```

Comment:
_oof,_ i know this is copied from Google's sample code, but let's use [urllib urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) instead